sklearn/utils/estimator_checks.py

```diff
@@ -385,55 +394,244 @@ def check_transformer_pickle(name, Transformer):
+                          cloned.partial_fit(X, y, classes), m2)
+
+    elif isinstance(estimator, RegressorMixin):
+        ret_val = estimator.partial_fit(X, y)
+        assert_same_model(X, estimator, cloned.partial_fit(X, y), m2)
+
     else:
-        alg.partial_fit(X, y)
+        ret_val = estimator.partial_fit(X)
+        assert_same_model(X, estimator, cloned.partial_fit(X), m2)
 
-    assert_raises(ValueError, alg.partial_fit, X[:, :-1], y)
+    # Check if returns self
+    assert_equal(ret_val, estimator, msg=m1)
 
 
-def check_clustering(name, Alg):
+def check_partial_fit_reset_when_fit(name, Estimator):
+    """Check if the estimators get reset when fit is applied after a 
+    partial fit.
+    """
+   
+    # Estimators are assumed to be reset when the transorm / predic result
+    # is similar to 1st fit call
+    if not hasattr(Estimator, 'partial_fit'):
+        return
+
+    X, y = make_blobs(n_samples=50, random_state=1)
+    X -= X.min()
+
+    with warnings.catch_warnings(record=True):
+        estimator = Estimator()
+   
+    set_random_state(estimator)
+    set_fast_parameters(estimator)
+    
+    classes = np.unique(y)
+
+    # Initial call to fit
+    if isinstance(estimator, (ClassifierMixin, RegressorMixin)):
+        estimator.fit(X, y)
+    else:
+        estimator.fit(X)
+    
+    # Save the model in current state (after one fit call)
+    model_after_first_fit = deepcopy(estimator)
+
+    # Run 2 iterations of full set of partial_fit with batches of size 10
+    for i in range(2):
+        X_s, y_s = shuffle(X, y, random_state=0)
+        for j in range(0, 50, 10):
+            if isinstance(estimator, ClassifierMixin):
+                estimator.partial_fit(X_s[j:10+j], y_s[j:10+j],
+                                      classes=classes)
+            elif isinstance(estimator, RegressorMixin):
+                estimator.partial_fit(X_s[j:10+j], y_s[j:10+j])
+            else:
+                estimator.partial_fit(X_s[j:10+j])
+   
+    # Make sure that the partial_fit model is not same as first fit model
+    # assert_not_same_model(X, estimator, model_after_first_fit) 
+
+    emsg = "Change in features during 2nd fit raised an error.\n" 
+
+    # Use fit(x, y) if classifier / regressor.
+    if isinstance(estimator, (ClassifierMixin, RegressorMixin)):
+        # Test that change in features does not raise any error
+        try:
+            estimator.fit(X[:, :-1], y)
+        except Exception as e:
+            # Change in features during the second fit call raised an error
+            assert False, emsg + str(e)
+        # Finally call fit to reset the estimator
+        estimator.fit(X, y)
+
+    # Use fit(x) otherwise.
+    else:
+        try:
+            estimator.fit(X[:, :-1])
+        except Exception as e:
+            assert False, emsg + str(e)
+        estimator.fit(X)
+
+    assert_same_model(X, estimator, model_after_first_fit)
+
+
+def check_partial_fit_effect(name, Estimator):
```

Comment:
On 01/17/2015 06:36 AM, jnothman wrote:
>
> And yes, it's possible that the test I describe above will also fail 
> for many estimators, but then we should understand why it's a poor 
> test and investigate how we might otherwise check the same property.
>
>
I think it is a good test :)